File: iteration2/src/CSE3063F22P1_GRP9/OutputJSON.java
package CSE3063F22P1_GRP9;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.Map;

import org.apache.log4j.Logger;
import org.json.simple.JSONArray;
import org.json.simple.JSONValue;

import com.fasterxml.jackson.databind.ObjectMapper;

public class OutputJSON {
	final Logger logger = Logger.getLogger(OutputJSON.class);
	public void saveStudent(Student student){
		Map map = new LinkedHashMap();
		map.put("ID", student.getID());
		map.put("FirstName",student.getFirstName());
		map.put("LastName",student.getLastName());
		map.put("Semester",student.getSemester()+1);
		map.put("Advisor",student.getAdvisor());
		map.put("Transcript",getTranscriptHashMap(student.getTranscript()));
		map.put("RequestedCourses",student.getRequestedCourses());
		try {
			ObjectMapper mapper = new ObjectMapper();
			String jsonString = mapper.writerWithDefaultPrettyPrinter().writeValueAsString(map);
			File dir = new File("students");
			dir.mkdir();
			FileWriter file = new FileWriter("students/%s.json".formatted(student.getID()));
			file.write(jsonString);
			file.close();
			logger.info("Student json file be created, file name is:" +student.getID()+".json");
		}catch(IOException e) {
			logger.error("Student json file couldn't be created");
			System.exit(3);
		}
	}
	
	private Map getTranscriptHashMap(Transcript transcript) {
		Map map = new LinkedHashMap();
		map.put("GPA", String.format("%.2f", transcript.getGPA()));
		map.put("TakenCredit", transcript.getTakenCredit());
		map.put("CompletedCredit",transcript.getCompletedCredit());
		map.put("TakenCourses",getJSONArrayofTakenCourses(transcript.getTakenCourses()));
		map.put("RegistirationProblems",getJSONArrayOfSelectionProblems(transcript.getSelectionProblems()));
		return map;
	}

	private Object getJSONArrayofTakenCourses(ArrayList<TakenCourse> takenCourses) {
		JSONArray takenCoursesMap = new JSONArray();
		for(int i = 0;i<takenCourses.size();i++) {
			Map map = new LinkedHashMap();
			TakenCourse takenCourse = takenCourses.get(i);
			map.put("CourseId",takenCourse.getCourse().getID());
			map.put("CourseName",takenCourse.getCourse().getName());
			map.put("Credit",takenCourse.getCourse().getCredit());
			map.put("takenCourseStatus",takenCourse.getTakenCourseStatus());
			map.put("Grade",takenCourse.getGrade());
			takenCoursesMap.add(map);
		}
		return takenCoursesMap;
	}

	private JSONArray getJSONArrayOfSelectionProblems(ArrayList<SelectionProblem> selectionProblems) {
		JSONArray selectionProblemsMap = new JSONArray();
		for(int i = 0;i<selectionProblems.size();i++) {
			Map map = new LinkedHashMap();
			SelectionProblem problem = selectionProblems.get(i);
			map.put("ProblemId",problem.getId());
			map.put("Course",problem.getNotRegisteredCourse().getName());
			map.put("Reason",problem.getDescription());
			selectionProblemsMap.add(map);
		}
		return selectionProblemsMap;
	}
	
	public void saveCourseStatistics(ArrayList<ArrayList<Course>> courses,String semester) {
		String allCourseStatistics = "";
		for(int i = 0;i<courses.size();i++) {
			for(int j = 0;j<courses.get(i).size();j++) {
				Course course = courses.get(i).get(j);
				if(!course.getSemester().equals(semester)) continue;
				CourseStatistics courseStatistics = course.getCourseStatistics();
				String s = course.getID()+" "+courses.get(i).get(j).getName()+" Statistics;\n";
				s += "	Percentage of sucessfull registerion:"+ Math.round(courseStatistics.getRateOfSuccessfulRegistration() * 100) +"%\n";
				s += "	Number of registered students:"+courseStatistics.getRegisteredStudentCount()+"\n";
				s += "	Number of registration failures:"+courseStatistics.getRegistrationFailureCount()+"\n";
				s += "	Number of quota problems:"+courseStatistics.getQuotaProblemCount()+"\n";
				s += "	Number of prerequisite problems:"+courseStatistics.getPrerequisiteProblemCount()+"\n";
			
				//System.out.print(s);
				logger.info(s);
				allCourseStatistics += s;
			}
		}
		try {
			FileWriter file = new FileWriter("DepartmentStatistics.txt");
			file.write(allCourseStatistics);
			file.close();
			logger.info("Department Statistics file is created.- DepartmentStatistics.txt");
		} catch (IOException e) {
			logger.error("Department Statistics couldn't be saved");
			System.exit(1);
		}
	}
}

File: iteration3/src/OutputJSON.py
from Student import Student
from Advisor import Advisor
from Transcript import Transcript
import json
import logging
import sys

class OutputJSON:

    def __init__(self):
        self._logger = logging.getLogger(__name__)

    def saveStudent(self,student:Student):
        student_dict = {}
        student_dict['ID'] = student.getID()
        student_dict['First Name'] = student.getFirstName()
        student_dict['Last Name'] = student.getLastName()
        student_dict['Semester'] = student.getSemester()
        student_dict['Advisor'] = self.getAdvisorDict(student.getAdvisor())
        student_dict['Transcript'] = self.getTranscriptDict(student.getTranscript())
        student_dict['Requested Courses'] = self.getRequstedCoursedict(student)
        json_object = json.dumps(student_dict, indent = 4,ensure_ascii=False)
        try:
         f = open(f'students/{student.getID()}.json', "w",encoding='utf-8')
        except:
         self._logger.warning('Unexpected error.')
         sys.exit()
        try: 
         f.write(json_object)
        except IOError as e :
         self._logger.warning("I/O error({0}): {1}".format(e.errno, e.strerror))
         sys.exit()
        f.close()
        self._logger.info(f"Student information for {student.getID()} saved successfully in students/{student.getID()}.json")
        
    

    def getAdvisorDict(self,advisor:Advisor):
        advisor_dict = {}
        advisor_dict['ID'] = advisor.getID()
        advisor_dict['First Name'] = advisor.getFirstName()
        advisor_dict['Last Name'] = advisor.getLastName()
        advisor_dict['Email'] = advisor.get_email()
        advisor_dict['Office'] = advisor.get_office()
        return advisor_dict

    def getTranscriptDict(self,tr:Transcript):
        tr_dict = {}
        tr_dict['GPA'] = tr.getGpa()
        tr_dict['Taken Credit'] = tr.getTakenCredit()
        tr_dict['Completed Credit'] = tr.getCompletedCredit()
        taken_courses_arr = []
        for taken_course in tr.getTakenCourses():
            taken_course_dict = {
                'Course ID' : taken_course.getCourse().getID(),
                'Course Name': taken_course.getCourse().getName(),
                'Credit': taken_course.getCourse().getCredit(),
                'Status': taken_course.getTakenCourseStatus(),
                'Grade': taken_course.getGrade()
            }
            taken_courses_arr.append(taken_course_dict)
        tr_dict['Taken Courses'] = taken_courses_arr
        registration_problems_arr = []
        for registration_problem in tr.getSelectionProblems():
            registration_problem_dict = {
                'Problem Id': registration_problem.get_id(),
                'Course' : registration_problem.get_not_registered_course().getName(),
                'Reason' : registration_problem.get_description()
            }
            registration_problems_arr.append(registration_problem_dict)
        tr_dict['Registration Problems'] = registration_problems_arr
        return tr_dict

    def getRequstedCoursedict(self,student:Student):
        rq = []
        for course in student.getRequestedCourses():
         rq_dict = {
            'ID':course.getID(),
            'Course Name': course.getName(),
            'Prerequisite Id': course.getPrerequisiteId(),
            'credit':course.getCredit(),
         }
         rq.append(rq_dict)
        return rq
       

    def saveCourseStatistics(self,courses,semester):
        allCourseStatistics = ''
        for sem in courses:
            for course in sem:
                if course.getSemester() != semester: continue
                allCourseStatistics += str(course.getCourseStatistics())
                print(allCourseStatistics)
        
        try:
         f = open(f'DepartmentStatistics.txt', "w",encoding='utf-8')
        except:
         self._logger.warning('Unexpected error.')
         sys.exit()
        try:
         f.write(allCourseStatistics)
        except IOError as e :
         self._logger.warning("I/O error({0}): {1}".format(e.errno, e.strerror))
         sys.exit()
        f.close()    
       
        
    
        
        